toolkit/tools/imagegen/diskutils/filesystem.go

```diff
@@ -0,0 +1,324 @@
+		"rmapbt":     "metadata",
+		"sparse":     "inode",
+		"nrext64":    "inode",
+	}
+
+	// The maximum version of mkfs.xfs that is currently supported.
+	// This is used to prevent issues with newer versions of mkfs.xfs default enabling new features.
+	maxMkfsXfsVersion = version.Version{6, 9}
+
+	// The minmum supported kernel version. This helps avoid versions complexity for features that are old and therefore
+	// basically universal.
+	//
+	// Relevant kernel versions:
+	// - Ubuntu 22.04: v5.15
+	// - Mariner 2.0: v5.15
+	// - Mariner 3.0: v6.6
+	minKernelVersion = version.Version{5, 4}
+
+	// For exampke: mke2fs 1.47.0 (5-Feb-2023)
+	mke2fsVersionRegex = regexp.MustCompile(`(?m)^mke2fs (\d+)\.(\d+)\.(\d+) \(\d+-[a-zA-Z]+-\d+\)$`)
+
+	// For example: mkfs.xfs version 6.5.0
+	mkfsXfsVersionRegex = regexp.MustCompile(`^mkfs\.xfs version (\d+)\.(\d+)\.(\d+)$`)
+)
+
+func getFileSystemOptions(targetOs targetos.TargetOs, filesystemType string) ([]string, error) {
+	hostKernelVersion, err := kernelversion.GetBuildHostKernelVersion()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get host kernel version:\n%w", err)
+	}
+
+	if minKernelVersion.Gt(hostKernelVersion) {
+		return nil, fmt.Errorf("host kernel version (%s) is too old (min: %s)", minKernelVersion, hostKernelVersion)
+	}
+
+	options, hasOptions := targetOsFileSystemsOptions[targetOs]
+	if !hasOptions {
+		return nil, fmt.Errorf("unknown target OS (%s)", targetOs)
+	}
+
+	switch filesystemType {
+	case "ext4":
+		options, err := getExt4FileSystemOptions(hostKernelVersion, options)
+		if err != nil {
+			return nil, err
+		}
+
+		return options, nil
+
+	case "xfs":
+		options, err := getXfsFileSystemOptions(hostKernelVersion, options)
+		if err != nil {
+			return nil, err
+		}
+
+		return options, nil
+
+	default:
+		return []string(nil), nil
+	}
+}
+
+func getExt4FileSystemOptions(hostKernelVersion version.Version, options fileSystemsOptions) ([]string, error) {
+	mke2fsVersion, err := getMke2fsVersion()
+	if err != nil {
+		return nil, err
+	}
+
+	// Omit features not supported by the build host kernel.
+	features := []string(nil)
+	for _, feature := range options.Ext4.Features {
+		requiredKernelVersion, hasRequiredKernelVersion := ext4FeaturesKernelSupport[feature]
+		if hasRequiredKernelVersion && requiredKernelVersion.Gt(hostKernelVersion) {
+			// Feature is not supported on build host kernel.
+			logger.Log.Infof("Build host kernel does not support ext4 feature (%s)", feature)
+			continue
+		}
+
+		requiredMke2fsVersion, hasRequiredMke2fsVersion := ext4FeaturesE2fsprogsSupport[feature]
+		if hasRequiredMke2fsVersion && requiredMke2fsVersion.Gt(mke2fsVersion) {
+			// Feature is not supported by version of mkfs.ext4.
+			logger.Log.Infof("mkfs.ext4 does not support ext4 feature (%s)", feature)
+			continue
+		}
+
+		features = append(features, feature)
+	}
+
+	// "none" requests no default options.
+	featuresArg := "none," + strings.Join(features, ",")
```

Comment:
Just checking: in the unlikely event that `features` is empty, will the trailing comma be an issue or will the tool just ignore it?

Reply:
Fixed.